### DISCUSSION TITLE
Carlo Gavazzi EM340 wired m-bus

### DISCUSSION THREAD

**Discussion by jakobschou** [wmbusmeters/wmbusmeters#842] [Q&A]
Hi,

I am considering buying the  energy meter Carlo Gavazzi EM340, which has a wired m-bus port (Protocol M-bus according to
EN13757-1), which I hope could be connected to Home Assistant with wmbusmeters using a USB to m-bus (like https://www.aliexpress.com/item/1005004533668296.html - slave version).

I cannot seem to find out if wmbusmeter has a driver for it...

Has anyone gotten this up and running?

**Comment by weetmuts** [comment]
The usb to mbus dongle should work.
Make sure you buy the master. You use the color or something in the alibaba ui.

Then you should be able to get mbus data from the meter and we can add a driver for it.

**Reply by jakobschou** [reply]
Thank you for the quick reply. 

I hope that can work - after connecting via the USB interface, should I just come back with logs or something? :-)

**Comment by weetmuts** [comment]
Yes assuming the meter has a secondary address configured already (like 12345678) and uses 2400 bps then you run.
wmbusmeters --pollinterval=60s --logtelegrams  MAIN=/dev/ttyUSB0:mbus:2400 MyMeter em24:MAIN:mbus 12345678 NOKEY

post the telegram=|......| lines here

**Reply by jakobschou** [reply]
Thank you!
Now I have gotten the meter installed, and are just waiting for the m-bus to usb master to arrive.

I will reply with telegram whenever possible! :-)

**Reply by jakobschou** [reply]
I have now received the mbus master, and connected it to my meter on my RPI4 (running HASSIO), whereas I got wmbusmeters installed.

Stupid question, but how do I get to run the command you suggested within Home Assistant terminal?

**Reply by jakobschou** [reply]
I am using the new HA addon (edge version), and have tried to set device= as "/dev/ttyUSB1:mbus:2400", "MAIN=/dev/ttyUSB1:mbus:2400", and just "/dev/ttyUSB1". But I don't know where to set the secondary address.

It does not seem to find the device when using "/dev/ttyUSB1:mbus:2400" and "/dev/ttyUSB1" when looking in logs:
```
INFO: Running wmbusmeters ...
No wmbus device detected, waiting for a device to be plugged in.
```

When using device as "MAIN=/dev/ttyUSB1:mbus:2400" I get this result:
```
INFO: Running wmbusmeters ...
Started config MAIN=mbus on /dev/ttyUSB1
```
But how I do set the secondary address for wired mbus?

Is this the correct way to use the HA addon with USB mbus devices? @BIBOLV

**Reply by BIBOLV** [reply]
`MAIN=/dev/ttyUSB1:mbus:2400;/dev/ttyUSB2`

**Reply by jakobschou** [reply]
> `MAIN=/dev/ttyUSB1:mbus:2400;/dev/ttyUSB2`

Thank you for the answer!
I am only using this one USB device, but is there a way to set the secondary address in HASSIO?

And is there a way to poll a meter using the previous mentioned command within HA, or can this only be done with a build from source?

**Comment by weetmuts** [comment]
If you read the manual you will see that you have to poll mbus meters.

wmbusmeters --pollinterval=60s MAIN=/dev/ttyUSB0:mbus:2400 MyMeter em24:MAIN:mbus 12345678 NOKEY

I do not know how to provide the pollinternal in the ha-addon though. It can be set in the MyMeter.conf file.
@BIBOLV can you help?

**Reply by BIBOLV** [reply]
I misunderstood about secondary address, I thought that secondary USB was needed 😉 But basically you make configuration same way it would be done using manual file config:
![image](https://user-images.githubusercontent.com/49994376/227462146-0b5824c8-a32e-4bbe-827b-f6f529f88adc.png)

**Reply by jakobschou** [reply]
Perfect!

That worked and I got the telegram😃

`(meter) EM340: meter detection did not match the selected driver em24! correct driver is: unknown!
(meter) Not printing this warning again for id: 00000000 mfct: (@@@) Unknown (0x00) type: Other (0x00) ver: 0x00
(meter) please consider opening an issue at https://github.com/wmbusmeters/wmbusmeters/
(meter) to add support for this unknown mfct,media,version combination
telegram=|6865656808007276301790361CC70200000000|0405EB12000004FB827501000000042AB400000004FB977212FCFFFF04FBB772FE03000002FDBA73B20084808040FD48A30F000004FD48070900008440FD59D3010000848040FD590000000084C040FD59000000001F|7116|+4
telegram=|6865656808007276301790361CC70201000000|0405EB12000004FB827501000000042AB400000004FB977210FCFFFF04FBB7720004000002FDBA73B00084808040FD48A10F000004FD48060900008440FD59D4010000848040FD590000000084C040FD59000000001F|6F16|+66
telegram=|6865656808007276301790361CC70202000000|0405EB12000004FB827501000000042AB300000004FB97720BFCFFFF04FBB7720504000002FDBA73AF0084808040FD48A80F000004FD480A0900008440FD59D5010000848040FD590000000084C040FD59000000001F|7A16|+128
telegram=|6865656808007276301790361CC70203000000|0405EB12000004FB827501000000042AB300000004FB97720AFCFFFF04FBB7720504000002FDBA73AE0084808040FD48A90F000004FD480B0900008440FD59D5010000848040FD590000000084C040FD59000000001F|7B16|+191
telegram=|6865656808007276301790361CC70204000000|0405EB12000004FB827501000000042AB400000004FB97720AFCFFFF04FBB7720604000002FDBA73AE0084808040FD48A60F000004FD48090900008440FD59D6010000848040FD590000000084C040FD59000000001F|7A16|+253`

**Comment by weetmuts** [comment]
Great! Can you run with log=debug just during a capture of  a single telegram?
I need to fix the mbus auto-detect problem here. :-)

**Reply by jakobschou** [reply]
I got this from log - can see it read total energy consumption.

`(em24) 004   : 08 dll-c (RSP_UD2)
(em24) 005   : 00 dll-a primary (0)
(em24) 006   : 72 tpl-ci (long header)
(em24) 007   : 76301790 tpl-id (90173076)
(em24) 011   : 361c tpl-mfct (GAV)
(em24) 013   : c7 tpl-version
(em24) 014   : 02 tpl-type (Electricity meter)
(em24) 015   : 17 tpl-acc-field
(em24) 016   : 00 tpl-sts-field (OK)
(em24) 017   : 0000 tpl-cfg 0000 ( )
(em24) 019   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 020   : 05 vif (Energy 10² Wh)
(em24) 021 C!: EB120000 ("total_energy_consumption_kwh":484.3)
(em24) 025   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 026   : FB vif (First extension FB of VIF-codes)
(em24) 027   : 82 vife (Reserved)
(em24) 028   : 75 combinable vif (MultiplicativeCorrectionFactorForValue)
(em24) 029 C!: 01000000 ("total_reactive_energy_consumption_kvarh":1)
(em24) 033   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 034   : 2A vif (Power 10⁻¹ W)
(em24) 035 C?: B4000000
(em24) 039   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 040   : FB vif (First extension FB of VIF-codes)
(em24) 041   : 97 vife (Reserved)
(em24) 042   : 72 combinable vif (MultiplicativeCorrectionFactorForValue)
(em24) 043 C?: 0DFCFFFF
(em24) 047   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 048   : FB vif (First extension FB of VIF-codes)
(em24) 049   : B7 vife (Reserved)
(em24) 050   : 72 combinable vif (MultiplicativeCorrectionFactorForValue)
(em24) 051 C?: 03040000
(em24) 055   : 02 dif (16 Bit Integer/Binary Instantaneous value)
(em24) 056   : FD vif (Second extension FD of VIF-codes)
(em24) 057   : BA vife (Dimensionless / no VIF)
(em24) 058   : 73 combinable vif (MultiplicativeCorrectionFactorForValue)
(em24) 059 C?: B000
(em24) 061   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 062   : 80 dife (subunit=0 tariff=0 storagenr=0)
(em24) 063   : 80 dife (subunit=0 tariff=0 storagenr=0)
(em24) 064   : 40 dife (subunit=4 tariff=0 storagenr=0)
(em24) 065   : FD vif (Second extension FD of VIF-codes)
(em24) 066   : 48 vife (10^-1 Volts)
(em24) 067 C?: A20F0000
(em24) 071   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 072   : FD vif (Second extension FD of VIF-codes)
(em24) 073   : 48 vife (10^-1 Volts)
(em24) 074 C?: 07090000
(em24) 078   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 079   : 40 dife (subunit=1 tariff=0 storagenr=0)
(em24) 080   : FD vif (Second extension FD of VIF-codes)
(em24) 081   : 59 vife (10^-3 Ampere)
(em24) 082 C?: D6010000
(em24) 086   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 087   : 80 dife (subunit=0 tariff=0 storagenr=0)
(em24) 088   : 40 dife (subunit=2 tariff=0 storagenr=0)
(em24) 089   : FD vif (Second extension FD of VIF-codes)
(em24) 090   : 59 vife (10^-3 Ampere)
(em24) 091 C?: 00000000
(em24) 095   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(em24) 096   : C0 dife (subunit=1 tariff=0 storagenr=0)
(em24) 097   : 40 dife (subunit=3 tariff=0 storagenr=0)
(em24) 098   : FD vif (Second extension FD of VIF-codes)
(em24) 099   : 59 vife (10^-3 Ampere)
(em24) 100 C?: 00000000
(em24) 104 C?: 1F unknown dif treating remaining data as mfct specific 
(meters) render field total_energy_consumption(Energy kwh)[2] with dventry @21 key 0405 data EB120000
(meters)             "total_energy_consumption_kwh":484.3
(meters) render field total_reactive_energy_consumption(Reactive_Energy kvarh)[4] with dventry @29 key 04FB8275 data 01000000
(meters)             "total_reactive_energy_consumption_kvarh":1
(meters) render field total_apparent_energy_consumption(Apparent_Energy)[6] without dventry
(meters)             "total_apparent_energy_consumption_kvah":484.301032
(meters) render field total_apparent_energy_production(Apparent_Energy)[7] without dventry
(meters)             "total_apparent_energy_production_kvah":null
(shell) exec "/bin/sh"
(shell) arg "-c"
(shell) arg "/wmbusmeters/mosquitto_pub.sh "wmbusmeters/$METER_NAME" "$METER_JSON""
(shell) env "METER_JSON={"media":"electricity","meter":"em24","name":"EM340","id":"90173076","total_energy_consumption_kwh":484.3,"total_reactive_energy_consumption_kvarh":1,"total_apparent_energy_consumption_kvah":484.301032,"total_apparent_energy_production_kvah":null,"timestamp":"2023-03-24T08:40:23Z","device":"MAIN","rssi_dbm":0}"
(shell) env "METER_ID=90173076"
(shell) env "METER_NAME=EM340"
(shell) env "METER_MEDIA=electricity"
(shell) env "METER_TYPE=em24"
(shell) env "METER_TIMESTAMP=2023-03-24T08:40:23Z"
(shell) env "METER_TIMESTAMP_UTC=2023-03-24T08:40:23Z"
(shell) env "METER_TIMESTAMP_UT=1679647223"
(shell) env "METER_TIMESTAMP_LT=2023-03-24 09:40.23"
(shell) env "METER_DEVICE=MAIN"
(shell) env "METER_RSSI_DBM=0"
(shell) env "METER_STATUS=null"
(shell) env "METER_ERROR=null"
(shell) env "METER_TOTAL_ENERGY_CONSUMPTION_KWH=484.3"
(shell) env "METER_TOTAL_ENERGY_PRODUCTION_KWH=null"
(shell) env "METER_TOTAL_REACTIVE_ENERGY_CONSUMPTION_KVARH=1"
(shell) env "METER_TOTAL_REACTIVE_ENERGY_PRODUCTION_KVARH=null"
(shell) env "METER_TOTAL_APPARENT_ENERGY_CONSUMPTION_KVAH=484.301032"
(shell) env "METER_TOTAL_APPARENT_ENERGY_PRODUCTION_KVAH=null"
(shell) waiting for child 254 to complete.
(shell) /bin/sh: return code 0
(mbus) checkMBUSFrame
 ""
(mbus) less than 6 bytes, partial frame`

**Comment by weetmuts** [comment]
loglevel=debug

**Comment by weetmuts** [comment]
Ok, you can see the decoding here:
https://wmbusmeters.org/analyze/6865656808007276301790361cc702000000000405eB12000004fB827501000000042aB400000004fB977212fcffff04fBB772fe03000002fdBa73B20084808040fd48a30f000004fd48070900008440fd59d3010000848040fd590000000084c040fd59000000001f7116:em24

It seems like can add some more fields to the driver?
Can you tell us what the other fields mean? You can also switch to the gransystems driver to compare.

**Reply by jakobschou** [reply]
Just tried the gransystems driver instead:

`(gransystems) 020   : 05 vif (Energy 10² Wh)
(gransystems) 021 C!: EB120000 ("total_energy_consumption_kwh":484.3)
(gransystems) 025   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 026   : FB vif (First extension FB of VIF-codes)
(gransystems) 027   : 82 vife (Reserved)
(gransystems) 028   : 75 combinable vif (MultiplicativeCorrectionFactorForValue)
(gransystems) 029 C?: 01000000
(gransystems) 033   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 034   : 2A vif (Power 10⁻¹ W)
(gransystems) 035 C?: B8000000
(gransystems) 039   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 040   : FB vif (First extension FB of VIF-codes)
(gransystems) 041   : 97 vife (Reserved)
(gransystems) 042   : 72 combinable vif (MultiplicativeCorrectionFactorForValue)
(gransystems) 043 C?: 02FCFFFF
(gransystems) 047   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 048   : FB vif (First extension FB of VIF-codes)
(gransystems) 049   : B7 vife (Reserved)
(gransystems) 050   : 72 combinable vif (MultiplicativeCorrectionFactorForValue)
(gransystems) 051 C?: 0F040000
(gransystems) 055   : 02 dif (16 Bit Integer/Binary Instantaneous value)
(gransystems) 056   : FD vif (Second extension FD of VIF-codes)
(gransystems) 057   : BA vife (Dimensionless / no VIF)
(gransystems) 058   : 73 combinable vif (MultiplicativeCorrectionFactorForValue)
(gransystems) 059 C?: B300
(gransystems) 061   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 062   : 80 dife (subunit=0 tariff=0 storagenr=0)
(gransystems) 063   : 80 dife (subunit=0 tariff=0 storagenr=0)
(gransystems) 064   : 40 dife (subunit=4 tariff=0 storagenr=0)
(gransystems) 065   : FD vif (Second extension FD of VIF-codes)
(gransystems) 066   : 48 vife (10^-1 Volts)
(gransystems) 067 C?: BD0F0000
(gransystems) 071   : 04 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 072   : FD vif (Second extension FD of VIF-codes)
(gransystems) 073   : 48 vife (10^-1 Volts)
(gransystems) 074 C!: 16090000 ("voltage_at_phase_1_v":232.6)
(gransystems) 078   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 079   : 40 dife (subunit=1 tariff=0 storagenr=0)
(gransystems) 080   : FD vif (Second extension FD of VIF-codes)
(gransystems) 081   : 59 vife (10^-3 Ampere)
(gransystems) 082 C!: D8010000 ("current_at_phase_1_a":0.472)
(gransystems) 086   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 087   : 80 dife (subunit=0 tariff=0 storagenr=0)
(gransystems) 088   : 40 dife (subunit=2 tariff=0 storagenr=0)
(gransystems) 089   : FD vif (Second extension FD of VIF-codes)
(gransystems) 090   : 59 vife (10^-3 Ampere)
(gransystems) 091 C!: 00000000 ("current_at_phase_2_a":0)
(gransystems) 095   : 84 dif (32 Bit Integer/Binary Instantaneous value)
(gransystems) 096   : C0 dife (subunit=1 tariff=0 storagenr=0)
(gransystems) 097   : 40 dife (subunit=3 tariff=0 storagenr=0)
(gransystems) 098   : FD vif (Second extension FD of VIF-codes)
(gransystems) 099   : 59 vife (10^-3 Ampere)
(gransystems) 100 C!: 00000000 ("current_at_phase_3_a":0)
(gransystems) 104 C?: 1F unknown dif treating remaining data as mfct specific 
(meters) render field status(Text)[0] without dventry
(meters)             "status":"OK"
(meters) render field total_energy_consumption(Energy kwh)[2] with dventry @21 key 0405 data EB120000
(meters)             "total_energy_consumption_kwh":484.3
(meters) render field voltage_at_phase_1(Voltage v)[8] with dventry @74 key 04FD48 data 16090000
(meters)             "voltage_at_phase_1_v":232.6
(meters) render field current_at_phase_{subunit_counter}(Amperage a)[11] with dventry @82 key 8440FD59 data D8010000
(meters)             "current_at_phase_1_a":0.472
(meters) render field current_at_phase_{subunit_counter}(Amperage a)[11] with dventry @91 key 848040FD59 data 00000000
(meters)             "current_at_phase_2_a":0
(meters) render field current_at_phase_{subunit_counter}(Amperage a)[11] with dventry @100 key 84C040FD59 data 00000000
(meters)             "current_at_phase_3_a":0
(meters) render field frequency(Frequency)[13] without dventry
(meters)             "frequency_hz":null
(shell) exec "/bin/sh"
(shell) arg "-c"
(shell) arg "/wmbusmeters/mosquitto_pub.sh "wmbusmeters/$METER_NAME" "$METER_JSON""
(shell) env "METER_JSON={"media":"electricity","meter":"gransystems","name":"EM340","id":"90173076","status":"OK","total_energy_consumption_kwh":484.3,"voltage_at_phase_1_v":232.6,"current_at_phase_1_a":0.472,"current_at_phase_2_a":0,"current_at_phase_3_a":0,"frequency_hz":null,"timestamp":"2023-03-24T08:52:22Z","device":"MAIN","rssi_dbm":0}"
(shell) env "METER_ID=90173076"
(shell) env "METER_NAME=EM340"
(shell) env "METER_MEDIA=electricity"
(shell) env "METER_TYPE=gransystems"
(shell) env "METER_TIMESTAMP=2023-03-24T08:52:22Z"
(shell) env "METER_TIMESTAMP_UTC=2023-03-24T08:52:22Z"
(shell) env "METER_TIMESTAMP_UT=1679647942"
(shell) env "METER_TIMESTAMP_LT=2023-03-24 09:52.22"
(shell) env "METER_DEVICE=MAIN"
(shell) env "METER_RSSI_DBM=0"
(shell) env "METER_STATUS=OK"
(shell) env "METER_INFO=null"
(shell) env "METER_TOTAL_ENERGY_CONSUMPTION_KWH=484.3"
(shell) env "METER_TOTAL_ENERGY_CONSUMPTION_TARIFF_{TARIFF_COUNTER}_KWH=null"
(shell) env "METER_TARGET_DateTimeLT=null"
(shell) env "METER_TARGET_ENERGY_CONSUMPTION_KWH=null"
(shell) env "METER_TARGET_ENERGY_CONSUMPTION_TARIFF_{TARIFF_COUNTER}_KWH=null"
(shell) env "METER_DEVICE_DateTimeLT=null"
(shell) env "METER_VOLTAGE_AT_PHASE_1_Volt=232.6"
(shell) env "METER_VOLTAGE_AT_PHASE_{SUBUNIT_COUNTER}_Volt=null"
(shell) env "METER_CURRENT_AT_PHASE_1_Ampere=0.472"
(shell) env "METER_CURRENT_AT_PHASE_{SUBUNIT_COUNTER}_Ampere=null"
(shell) env "METER_FREQUENCY_HZ=null"
(shell) waiting for child 249 to complete.
(shell) /bin/sh: return code 0
(mbus) checkMBUSFrame
 ""
(mbus) less than 6 bytes, partial frame`

**Reply by jakobschou** [reply]
Telegram:

`telegram=|6865656808007276301790361CC70228000000|0405EB12000004FB827501000000042AB700000004FB97720CFCFFFF04FBB7720404000002FDBA73B20084808040FD48B20F000004FD48100900008440FD59D5010000848040FD590000000084C040FD59000000001F|B716|+4`

**Reply by jakobschou** [reply]
I have tried to distinguish some of them:

We are missing:
voltage_at_phase_2_v
voltage_at_phase_3_v
current_power_consumption_kvar
frequency_hz

Maybe these:
total_apparent_energy_production_kvah
total_apparent_energy_production_kw

021 C!: EB120000 ("total_energy_consumption_kwh":484.3)
029 C!: 01000000 ("total_reactive_energy_consumption_kvarh":1)
035 C!: B4000000 ("current_power_consumption_kw":0.018)
043 C?: 12FCFFFF
051 C?: FE030000
059 C?: B200
067 C?: A30F0000
074 C!: 07090000 ("voltage_at_phase_1_v":231.1)
082 C!: D3010000 ("current_at_phase_1_a":0.467)
091 C!: 00000000 ("current_at_phase_2_a":0)
100 C!: 00000000 ("current_at_phase_3_a":0)
104 C?: 1F unknown dif treating remaining data as mfct specific

**Comment by jakobschou** [comment]
Do you need any more info from me or should I open an "Issue" instead? @weetmuts

**Comment by jakobschou** [comment]
Update:
For now I have started using the amiplus driver, since I am getting the 2 most important fields from the telegram; "METER_TOTAL_ENERGY_CONSUMPTION_KWH=659.5"
"METER_CURRENT_POWER_CONSUMPTION_KW=0.0194"

I have configured Home Assistant to manually listen for these 2 parameters over MQTT, and I can follow my energy usage.


When going through all the over 3 drivers in the analyze page, I made the list of fields I could find, but I don't know the rest.

If you need anything regarding for implementation, eg. YAML for Home Assistant - please let me know :-)

**Comment by thoath2** [comment]
Hello,
i'm new here. See i correct it, that you use the wmbusmeters-ha-addon for wired M-Bus-devices. Or how you did it?
Thomas

**Reply by jakobschou** [reply]
Hi, yes - I am using a cheap USB to M-bus (master) device directly connected to my HA-server to measure energy usage on a M-bus power meter.

However, I could only get the current power usage in KW and total energy consumption in kWh (which for me was enough)

If one wanted to address all the M-bus frames, you probably could read more data, but that will require a new driver for the specific meter. 

There is a website where you can test which driver works best with the read data from the meter.

**Reply by thoath2** [reply]
Thanks for your answer,
first of all, I have more fundamental questions. Which interface/addon you use in HA, to read it out from your  wired USB to M-bus (master) device?

**Comment by thoath2** [comment]
Hello I need help.
I try to read data from two wired MBUS-Devices.
Here are my device-Parameters. I use an USB-Stick (USB to MBUS (Master). The M-Bus-devices are connected via a wire.
In Windows 10 i could read all data with the Software M-Tool 4.7.exe. Its automaticly find the right devices and shows me all data.
Now i want to use it in HAOS. I installed in HA the addon "[edge] Wmbusmeters (W-MBus to MQTT)".
Now I don't know what to do.

The first device have the parameters:

- Device-Type: Wärmemengenzähler Sensostar U from Engelmann
- USB-Port: /dev/serial/by-id/usb-Prolific_Technology_Inc._USB-Serial_Controller_BECUn114J20-if00-port0
- Baudrate: 2400
- primary-mbus-adress: 0
- secondary-mbus-adress: 04352936
- databits: 8E1

The second device have the parameters:

- Device-Type: Wärmemengenzähler Sensostar U from Engelmann
- USB-Port: /dev/serial/by-id/usb-Prolific_Technology_Inc._USB-Serial_Controller_BECUn114J20-if00-port0
- Baudrate: 2400
- primary-mbus-adress: 0
- secondary-mbus-adress: 04352932
- databits: 8E1

I want to read the data in 60s interval.

Is there a step-by-step manual to configure. With the documentation i'm not able to do it.

I look forward to your answers.
Thomas

**Reply by thoath2** [reply]
Juhu, it works. 
Thanks to the screenshot of Jakobschou and the documentation, I got the access of two wired M-bus devices!
I took the following steps:

- Connected to the meters on Windows PC with gateway.

- Using the M-Tools 4.7.exe software to learn the telegram and secondary addresses.

- On the page https://wmbusmeters.org I have entered the telegram
`68 B3 B3 68 08 00 72 36 29 35 04 C5 14 00 04 70 00 00 00 04 78 A8 6B 42 00 04 06 00 00 00 00 04 13 43 00 00 00 04 2B 00 00 00 00 14 2B 2F 01 00 00 04 3B 00 00 00 00 14 3B FD 00 00 00 02 5B 15 00 02 5F 15 00 02 61 FB FF 02 23 A1 05 04 6D 05 26 01 3A 44 06 00 00 00 00 44 13 00 00 00 00 42 6C FF 2C 01 FD 17 00 03 FD 0C 05 00 00 84 20 06 00 00 00 00 C4 20 06 00 00 00 00 84 30 06 00 00 00 00 C4 30 06 00 00 00 00 84 40 13 00 00 00 00 C4 40 13 00 00 00 00 84 80 40 13 00 00 00 00 C4 80 40 13 00 00 00 00 84 C0 40 13 00 00 00 00 C4 C0 40 13 00 00 00 00 D6 16`

- There I learned that the driver "**sensostar**" fits.

- Installation of the HA-Addon "[Edge] Wmbusmeters (W-BUS to MQTT)" https://github.com/wmbusmeters/wmbusmeters-ha-addon. This was a bit hooky, since 2 addons were installed during the installation and did not know which one I should take. I then took the Edge ...

- Then via the sidebar for configuration. I configured there:

<img width="794" alt="Konfiguration" src="https://github.com/user-attachments/assets/a29d9eb5-bb22-4024-9469-9836fdec664c">

- Device: MAIN=/dev/serial/by-id/usb-Prolific_Technology_Inc._USB-Serial_Controller_BECUn114J20-if00-port0:mbus:2400 (to find out the port I used the following Statement at terminal):
`ls -lR /dev/ | grep -v '\->\s../tty' | grep -v ttyprintk | grep -e 'tty[[:alpha:]]' -e serial`

- Format: JSON

- And then I created two meters:

- meter 1:

driver: sensostar:MAIN:mbus
id: 04352936 -> This is the secondary address of the first counter
key: NOKEY
name: test_wmz_links
pollinterval: 60s -> newly created

- meter 2:

driver: sensostar:MAIN:mbus
id: 04352932 -> This is the secondary address of the second counter
key: NOKEY
name: test_wmz_rechts
pollinterval: 60s -> newly created

- Then I saved and restarted ha.

- And a miracle happened. I could see the success in the log and what was even better, the addon has already created the sensors for MQTT without doing anything in the configuration.yaml.

**Now I have two questions to You:**

1. A question is derived from this. So it is no longer necessary, as described in the documentation, to write in configuration.yaml like this:
`mqtt:
  sensor:
    - state_topic: "wmbusmeters/MainWater"
      json_attributes_topic: "wmbusmeters/MainWater"
      unit_of_measurement: m³
      value_template: "{{ value_json.total_m3 }}"
      name: Water usage
      icon: "mdi:gauge"
      state_class: total_increasing
      device_class: water`

2. What would I have to enter in the configuration area "wmbusmeters configuration" at device:
`MAIN=/dev/serial/by-id/usb-Prolific_Technology_Inc._USB-Serial_Controller_BECUn114J20-if00-port0:mbus:2400 `
if I would use two USB gateways?

**Reply by BIBOLV** [reply]
MAIN=/dev/ttyUSB1:mbus:2400;/dev/ttyUSB2:mbus:2400

**Reply by thoath2** [reply]
> MAIN=/dev/ttyUSB1:mbus:2400;/dev/ttyUSB2:mbus:2400

Thanks BIBOLV, i will try it, when i get the second device.
Thomas

**Comment by tuomaz** [comment]
Hi! I have a Gavazzi EM340 ([this one](https://gavazzi.se/produkter/em340dinav23xm1pfa/)) connected to a computer over a USB-MBUS device. It works together with wmbusmeters.

Today I get total increasing of consumed kWh which is good. I think it should be possible to parse even more data from the device, [this](https://gavazzi.se/app/uploads/2020/11/em330_em340_m1_cp.pdf) document suggest that. What can I do to improve the parsing?

Log from wmbusmeters:
[wmbusmeters-5k.log](https://github.com/user-attachments/files/17548129/wmbusmeters-5k.log)

Log from mbus-serial-request-data:
[mbus-serial-request-data.log](https://github.com/user-attachments/files/17548142/mbus-serial-request-data.log)

**Reply by tuomaz** [reply]
Logged telegrams:

6865656808007228720490361CC70251000000_04059A34050004FB827503C80000042A333E000004FB9772AD0A000004FBB772213F000002FDBA73DA0384808040FD48A30F000004FD48070900008440FD59E60B0000848040FD597608000084C040FD59630800001F_3B16

6878786808007228720490361CC70252000000_84402AE31900008480402A2412000084C0402A341200008440FB9772B7030000848040FB9772A803000084C040FB9772590300008440FBB772271A0000848040FBB7728212000084C040FBB772821200008240FDBA73DE03828040FDBA73D40382C040FDBA73D7031F_0616

6875756808007228720490361CC70253000000_84C08040FD48950F00008480C040FD48C30F000084C0C040FD48950F00008440FD48FC080000848040FD480309000084C040FD481A09000084808040059A34050084808040FB827503C8000084C08040050000000084C08040FB82750000000002FB2EF4011F_3416

6839396808007228720490361CC70254000000_844005DC060200848040052497010084C040050C970100848080402AF03C000084C080402AE03B02001F_7A16

6822226808007228720490361CC70255000000_8480C040050000000084C0C04005000000000F_7916

I think these belong together.

**Reply by tuomaz** [reply]
Got some more data with the gransystems driver, but I think there is even more to collect.
